{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9a4c923b-cc45-45b4-ac44-6e575e3d760f",
   "metadata": {},
   "source": [
    "# Alignment\n",
    "Some interesting and important info from lecture slides:\n",
    "3 RNA-seq mapping/alignment strategies\n",
    "1. de novo assembly: if the reference genome doesn't exit. Tools: Trinity, Velvet\n",
    "2. align to transcriptome: relies on known transcripts\n",
    "3. align to reference genome: allows for novel transcript discovery\n",
    "\n",
    "Genomic cordinates\n",
    "1-based: GFF, SAM, VCF, Ensembl browser, etc.\n",
    "0-based: BED, BAM, UCSC browser\n",
    "\n",
    "Alignment QC: https://github.com/griffithlab/rnabio.org/blob/master/assets/lectures/cshl/2024/mini/RNASeq_MiniLecture_02_04_alignmentQC.pdf"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ffd33e7-7422-4827-8171-9c7772a654d8",
   "metadata": {},
   "source": [
    "# Adapter trim\n",
    "Use Fastp to trim sequence adapter from the read FASTQ files and also perform basic data quality cleanup. The output of this step will be trimmed and filtered FASTQ files for each data set."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5baaee7e-79a7-49d1-b672-6048c74ca290",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/workspace/rnaseq/data/trimmed\n",
      "--2025-04-27 01:32:11--  http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa\n",
      "Resolving genomedata.org (genomedata.org)... 54.71.55.4\n",
      "Connecting to genomedata.org (genomedata.org)|54.71.55.4|:80... connected.\n",
      "HTTP request sent, awaiting response... 301 Moved Permanently\n",
      "Location: https://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa [following]\n",
      "--2025-04-27 01:32:11--  https://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa\n",
      "Connecting to genomedata.org (genomedata.org)|54.71.55.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 161\n",
      "Saving to: â€˜illumina_multiplex.faâ€™\n",
      "\n",
      "illumina_multiplex. 100%[===================>]     161  --.-KB/s    in 0s      \n",
      "\n",
      "2025-04-27 01:32:12 (134 MB/s) - â€˜illumina_multiplex.faâ€™ saved [161/161]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "echo $RNA_DATA_TRIM_DIR\n",
    "mkdir -p $RNA_DATA_TRIM_DIR\n",
    "cd $RNA_REFS_DIR\n",
    "\n",
    "#Download necessary Illumina adapter sequence files.\n",
    "wget http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db16e41b-0dcc-43ad-99f7-d28a8c72faec",
   "metadata": {},
   "source": [
    "Use fastp to remove illumina adapter sequences (if any), trim the first 13 bases of each read, and perform default read quality filtering to remove reads that are too short, have too many low quality bases or have too many Nâ€™s.\n",
    "\n",
    "* -l 25: the minimum read length allowed after trimming is 25bp\n",
    "* â€“adapter_fasta: the path to the adapter FASTA file containing adapter sequences to trim\n",
    "* â€“trim_front1 13: trim a fixed number (13 in this case) of bases off the left end of read1\n",
    "* â€“trim_front2 13: trim a fixed number (13 in this case) of bases off the left end of read2\n",
    "* â€“json: the path to store a log file in JSON file format\n",
    "* â€“html: the path to store a web report file\n",
    "* 2>: use to store the information that would be printed to the screen into a file instead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "de66d912-42f6-4333-a36a-02c68392961d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $RNA_HOME\n",
    "\n",
    "export S1=UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22\n",
    "fastp -i $RNA_DATA_DIR/$S1.read1.fastq.gz -I $RNA_DATA_DIR/$S1.read2.fastq.gz -o $RNA_DATA_TRIM_DIR/$S1.read1.fastq.gz -O $RNA_DATA_TRIM_DIR/$S1.read2.fastq.gz -l 25 --adapter_fasta $RNA_REFS_DIR/illumina_multiplex.fa --trim_front1 13 --trim_front2 13 --json $RNA_DATA_TRIM_DIR/$S1.fastp.json --html $RNA_DATA_TRIM_DIR/$S1.fastp.html 2>$RNA_DATA_TRIM_DIR/$S1.fastp.log\n",
    "\n",
    "export S2=UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22\n",
    "fastp -i $RNA_DATA_DIR/$S2.read1.fastq.gz -I $RNA_DATA_DIR/$S2.read2.fastq.gz -o $RNA_DATA_TRIM_DIR/$S2.read1.fastq.gz -O $RNA_DATA_TRIM_DIR/$S2.read2.fastq.gz -l 25 --adapter_fasta $RNA_REFS_DIR/illumina_multiplex.fa --trim_front1 13 --trim_front2 13 --json $RNA_DATA_TRIM_DIR/$S2.fastp.json --html $RNA_DATA_TRIM_DIR/$S2.fastp.html 2>$RNA_DATA_TRIM_DIR/$S2.fastp.log\n",
    "\n",
    "export S3=UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22\n",
    "fastp -i $RNA_DATA_DIR/$S3.read1.fastq.gz -I $RNA_DATA_DIR/$S3.read2.fastq.gz -o $RNA_DATA_TRIM_DIR/$S3.read1.fastq.gz -O $RNA_DATA_TRIM_DIR/$S3.read2.fastq.gz -l 25 --adapter_fasta $RNA_REFS_DIR/illumina_multiplex.fa --trim_front1 13 --trim_front2 13 --json $RNA_DATA_TRIM_DIR/$S3.fastp.json --html $RNA_DATA_TRIM_DIR/$S3.fastp.html 2>$RNA_DATA_TRIM_DIR/$S3.fastp.log\n",
    "\n",
    "export S4=HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22\n",
    "fastp -i $RNA_DATA_DIR/$S4.read1.fastq.gz -I $RNA_DATA_DIR/$S4.read2.fastq.gz -o $RNA_DATA_TRIM_DIR/$S4.read1.fastq.gz -O $RNA_DATA_TRIM_DIR/$S4.read2.fastq.gz -l 25 --adapter_fasta $RNA_REFS_DIR/illumina_multiplex.fa --trim_front1 13 --trim_front2 13 --json $RNA_DATA_TRIM_DIR/$S4.fastp.json --html $RNA_DATA_TRIM_DIR/$S4.fastp.html 2>$RNA_DATA_TRIM_DIR/$S4.fastp.log\n",
    "\n",
    "export S5=HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22\n",
    "fastp -i $RNA_DATA_DIR/$S5.read1.fastq.gz -I $RNA_DATA_DIR/$S5.read2.fastq.gz -o $RNA_DATA_TRIM_DIR/$S5.read1.fastq.gz -O $RNA_DATA_TRIM_DIR/$S5.read2.fastq.gz -l 25 --adapter_fasta $RNA_REFS_DIR/illumina_multiplex.fa --trim_front1 13 --trim_front2 13 --json $RNA_DATA_TRIM_DIR/$S5.fastp.json --html $RNA_DATA_TRIM_DIR/$S5.fastp.html 2>$RNA_DATA_TRIM_DIR/$S5.fastp.log\n",
    "\n",
    "export S6=HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22\n",
    "fastp -i $RNA_DATA_DIR/$S6.read1.fastq.gz -I $RNA_DATA_DIR/$S6.read2.fastq.gz -o $RNA_DATA_TRIM_DIR/$S6.read1.fastq.gz -O $RNA_DATA_TRIM_DIR/$S6.read2.fastq.gz -l 25 --adapter_fasta $RNA_REFS_DIR/illumina_multiplex.fa --trim_front1 13 --trim_front2 13 --json $RNA_DATA_TRIM_DIR/$S6.fastp.json --html $RNA_DATA_TRIM_DIR/$S6.fastp.html 2>$RNA_DATA_TRIM_DIR/$S6.fastp.log"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e5118f8e-e9bc-4f1d-afae-eb0683e48a01",
   "metadata": {},
   "source": [
    "## Use FastQC and multiqc to compare the impact of trimming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5072b8df-a5b4-4749-aa55-36097938fa2b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started analysis of HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "\n",
      "\u001b[38;5;208m///\u001b[0m \u001b]8;id=199630;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ ðŸ” \u001b[2mv1.28\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /workspace/rnaseq/data/trimmed\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[38;5;70mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m54/54\u001b[0m  /54\u001b[0m  \n",
      "\u001b[?25h\u001b[34m             fastp\u001b[0m | Found 6 reports\n",
      "\u001b[34m            fastqc\u001b[0m | Found 12 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m     write_results\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_DATA_TRIM_DIR\n",
    "fastqc *.fastq.gz\n",
    "multiqc ./"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43ea0d95-8e74-460e-8dec-b8c977af4c12",
   "metadata": {},
   "source": [
    "The resulting html reports can be viewed by navigating to:\n",
    "* http://YOUR_PUBLIC_IPv4_ADDRESS/rnaseq/data/\n",
    "* http://YOUR_PUBLIC_IPv4_ADDRESS/rnaseq/data/trimmed/\n",
    "\n",
    "* http://54.197.176.236/rnaseq/data/fastqc/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
    "* http://54.197.176.236/rnaseq/data/trimmed/fastqc/HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
    "* http://54.197.176.236/rnaseq/data/multiqc_report.html\n",
    "* http://54.197.176.236/rnaseq/data/trimmed/multiqc_report.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75f9c839-4e0d-4228-86f8-b870496ccdc5",
   "metadata": {},
   "source": [
    "## Move the fastqc and fastp results of trimmed file into sub-directories to keep things tidy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0f65f530-04e6-443c-a6fb-7319234111f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $RNA_DATA_TRIM_DIR\n",
    "mkdir fastqc\n",
    "mv *_fastqc* fastqc\n",
    "mkdir fastp\n",
    "mv *fastp.* fastp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b1850fc-b68c-4242-b24d-5579f4a14ab5",
   "metadata": {},
   "source": [
    "# PRACTICAL EXERCISE 5\n",
    "Using the approach above, trim the reads for both normal and tumor samples that you downloaded for the previous practical exercise. NOTE: try dropping the hard left trim option used above (â€˜â€“trim_front1 13â€™ and â€˜â€“trim_front2 13â€™). Once you have trimmed the reads, compare a pre- and post- trimming FastQ file using the FastQC and multiqc tools."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4bdbbccf-a3f0-46ad-a902-08680bfaf0e5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2025-04-27 01:58:18--  http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa\n",
      "Resolving genomedata.org (genomedata.org)... 54.71.55.4\n",
      "Connecting to genomedata.org (genomedata.org)|54.71.55.4|:80... connected.\n",
      "HTTP request sent, awaiting response... 301 Moved Permanently\n",
      "Location: https://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa [following]\n",
      "--2025-04-27 01:58:18--  https://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa\n",
      "Connecting to genomedata.org (genomedata.org)|54.71.55.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 161\n",
      "Saving to: â€˜illumina_multiplex.faâ€™\n",
      "\n",
      "illumina_multiplex. 100%[===================>]     161  --.-KB/s    in 0s      \n",
      "\n",
      "2025-04-27 01:58:19 (113 MB/s) - â€˜illumina_multiplex.faâ€™ saved [161/161]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_HOME/practice/data/\n",
    "mkdir trimmed\n",
    "wget http://genomedata.org/rnaseq-tutorial/illumina_multiplex.fa\n",
    "\n",
    "fastp -i hcc1395_normal_rep1_r1.fastq.gz -I hcc1395_normal_rep1_r2.fastq.gz -o trimmed/hcc1395_normal_rep1_r1.fastq.gz -O trimmed/hcc1395_normal_rep1_r2.fastq.gz -l 25 --adapter_fasta illumina_multiplex.fa --json trimmed/hcc1395_normal_rep1.fastp.json --html trimmed/hcc1395_normal_rep1.fastp.html 2>trimmed/hcc1395_normal_rep1.fastp.log\n",
    "fastp -i hcc1395_normal_rep2_r1.fastq.gz -I hcc1395_normal_rep2_r2.fastq.gz -o trimmed/hcc1395_normal_rep2_r1.fastq.gz -O trimmed/hcc1395_normal_rep2_r2.fastq.gz -l 25 --adapter_fasta illumina_multiplex.fa --json trimmed/hcc1395_normal_rep2.fastp.json --html trimmed/hcc1395_normal_rep2.fastp.html 2>trimmed/hcc1395_normal_rep2.fastp.log\n",
    "fastp -i hcc1395_normal_rep3_r1.fastq.gz -I hcc1395_normal_rep3_r2.fastq.gz -o trimmed/hcc1395_normal_rep3_r1.fastq.gz -O trimmed/hcc1395_normal_rep3_r2.fastq.gz -l 25 --adapter_fasta illumina_multiplex.fa --json trimmed/hcc1395_normal_rep3.fastp.json --html trimmed/hcc1395_normal_rep3.fastp.html 2>trimmed/hcc1395_normal_rep3.fastp.log\n",
    "\n",
    "fastp -i hcc1395_tumor_rep1_r1.fastq.gz -I hcc1395_tumor_rep1_r2.fastq.gz -o trimmed/hcc1395_tumor_rep1_r1.fastq.gz -O trimmed/hcc1395_tumor_rep1_r2.fastq.gz -l 25 --adapter_fasta illumina_multiplex.fa --json trimmed/hcc1395_tumor_rep1.fastp.json --html trimmed/hcc1395_tumor_rep1.fastp.html 2>trimmed/hcc1395_tumor_rep1.fastp.log\n",
    "fastp -i hcc1395_tumor_rep2_r1.fastq.gz -I hcc1395_tumor_rep2_r2.fastq.gz -o trimmed/hcc1395_tumor_rep2_r1.fastq.gz -O trimmed/hcc1395_tumor_rep2_r2.fastq.gz -l 25 --adapter_fasta illumina_multiplex.fa --json trimmed/hcc1395_tumor_rep2.fastp.json --html trimmed/hcc1395_tumor_rep2.fastp.html 2>trimmed/hcc1395_tumor_rep2.fastp.log\n",
    "fastp -i hcc1395_tumor_rep3_r1.fastq.gz -I hcc1395_tumor_rep3_r2.fastq.gz -o trimmed/hcc1395_tumor_rep3_r1.fastq.gz -O trimmed/hcc1395_tumor_rep3_r2.fastq.gz -l 25 --adapter_fasta illumina_multiplex.fa --json trimmed/hcc1395_tumor_rep3.fastp.json --html trimmed/hcc1395_tumor_rep3.fastp.html 2>trimmed/hcc1395_tumor_rep3.fastp.log"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d4a245a2-152d-488e-b4da-6b71b6c4cfef",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started analysis of hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "\n",
      "\u001b[38;5;208m///\u001b[0m \u001b]8;id=982255;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ ðŸ” \u001b[2mv1.28\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /workspace/rnaseq/practice/data/trimmed\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[38;5;70mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m54/54\u001b[0m  32m41/54\u001b[0m \u001b[2mhcc1395_tumor_rep3_r1_fastqc.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m             fastp\u001b[0m | Found 6 reports\n",
      "\u001b[34m            fastqc\u001b[0m | Found 12 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m     write_results\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_HOME/practice/data/trimmed/\n",
    "fastqc *.fastq.gz\n",
    "multiqc ./"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67e81f74-15cd-45a9-a601-6e6c53fcb862",
   "metadata": {},
   "source": [
    "* http://YOUR_IPV4/rnaseq/practice/data/hcc1395_normal_rep1_r1_fastqc.html\n",
    "* http://YOUR_IPV4/workspace/rnaseq/practice/data/trimmed/hcc1395_normal_rep1_1_fastqc.html\n",
    "* http://54.197.176.236/rnaseq/practice/data/hcc1395_normal_rep1_r1_fastqc.html\n",
    "* http://54.197.176.236/rnaseq/practice/data/trimmed/hcc1395_normal_rep1_r1_fastqc.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a849e11-c837-4f91-87ae-0b770eda1a40",
   "metadata": {},
   "source": [
    "1. After trimming, what is the range of read lengths observed for hcc1395 normal replicate 1, read 1?\n",
    "2. Which sections of the FastQC report are most informative for observing the effect of trimming?\n",
    "Ans: Basic Statisticsâ€™, â€˜Sequence Length Distributionâ€™, and â€˜Adapter Contentâ€™\n",
    "3. In the â€˜Per base sequence content sectionâ€™, what pattern do you see? What could explain this pattern?\n",
    "Ans: The first 9 base positions show a spiky pattern, suggesting biased representation of each base near the beginning of our reads/fragments. One possible explanation is that random hexamer priming for cDNA synthesis during library prep is happening in a non-random way. i.e. certain random hexamers are favored, therefore the creation of fragments (and ultimately reads) has a non-random pattern near the beginning."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f50626b-0031-4d8a-9791-e3115bbdea77",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd $RNA_HOME/practice/data\n",
    "mkdir fastqc\n",
    "mv *_fastqc* fastqc"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
