{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a7557f41-4d37-4397-9339-b48b4162f8a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;31mHBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mHBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mHBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mHBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mHBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mHBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mHBR_UHR_ERCC_ds_5pc.tar\u001b[0m\n",
      "\u001b[01;31mUHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mUHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mUHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mUHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\u001b[0m\n",
      "\u001b[01;31mUHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\u001b[0m\n",
      "\u001b[01;31mUHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_HOME/data\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd5c0ab6-b50c-4eb9-b0fd-21cbbec9b89f",
   "metadata": {},
   "source": [
    "# FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "26994c89-f3ff-416c-80e9-ef583f85b846",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started analysis of HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Started analysis of UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 5% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 10% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 15% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 20% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 25% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 30% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 35% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 40% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 45% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 50% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 55% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 60% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 65% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 70% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 75% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 80% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 85% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 90% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Approx 95% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Analysis complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1.fastq.gz\n",
      "Started analysis of UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 5% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 10% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 15% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 20% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 25% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 30% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 35% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 40% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 45% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 50% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 55% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 60% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 65% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 70% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 75% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 80% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 85% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 90% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Approx 95% complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n",
      "Analysis complete for UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "fastqc *.fastq.gz # * represents any character/word"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee709077-6895-41b8-9263-8005d8d8a4d4",
   "metadata": {},
   "source": [
    "* Then, go to the following url in your browser\n",
    "* Click on any of the *_fastqc.html files to view the FastQC report"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c98e0f-dd83-4763-bde2-52eefa100607",
   "metadata": {},
   "source": [
    "# MultiQC\n",
    "Run MultiQC on your fastqc reports to generate a single summary report across all samples/replicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "739ae5d9-c9c3-4d79-bdd0-9e828ee18ce7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\u001b[38;5;208m///\u001b[0m \u001b]8;id=784508;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2mv1.28\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /workspace/rnaseq/data\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[38;5;70m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m38/38\u001b[0m  /38\u001b[0m  \n",
      "\u001b[?25h\u001b[34m            fastqc\u001b[0m | Found 12 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m     write_results\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_HOME/data #not neccesary if you're already here\n",
    "multiqc ./"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49caa7de-3bcd-410a-b2c9-caebcdbc738f",
   "metadata": {},
   "source": [
    "* Then, go to the following url in your browser: http://YOUR_PUBLIC_IPv4_ADDRESS/rnaseq/data/multiqc_report.html"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "585f65b3-dc81-44c0-b614-7e4b3330a7c2",
   "metadata": {},
   "source": [
    "# Clean up\n",
    "Move all the FASTQC files into their own directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6b7b662f-2f53-408d-b5ed-db553ae8bd81",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
      "\u001b[0m\u001b[01;31mHBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.zip\u001b[0m\n",
      "HBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2_fastqc.html\n",
      "\u001b[01;31mHBR_Rep1_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2_fastqc.zip\u001b[0m\n",
      "HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
      "\u001b[01;31mHBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.zip\u001b[0m\n",
      "HBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2_fastqc.html\n",
      "\u001b[01;31mHBR_Rep2_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2_fastqc.zip\u001b[0m\n",
      "HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
      "\u001b[01;31mHBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read1_fastqc.zip\u001b[0m\n",
      "HBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2_fastqc.html\n",
      "\u001b[01;31mHBR_Rep3_ERCC-Mix2_Build37-ErccTranscripts-chr22.read2_fastqc.zip\u001b[0m\n",
      "UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
      "\u001b[01;31mUHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1_fastqc.zip\u001b[0m\n",
      "UHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2_fastqc.html\n",
      "\u001b[01;31mUHR_Rep1_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2_fastqc.zip\u001b[0m\n",
      "UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
      "\u001b[01;31mUHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1_fastqc.zip\u001b[0m\n",
      "UHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2_fastqc.html\n",
      "\u001b[01;31mUHR_Rep2_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2_fastqc.zip\u001b[0m\n",
      "UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1_fastqc.html\n",
      "\u001b[01;31mUHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read1_fastqc.zip\u001b[0m\n",
      "UHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2_fastqc.html\n",
      "\u001b[01;31mUHR_Rep3_ERCC-Mix1_Build37-ErccTranscripts-chr22.read2_fastqc.zip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_HOME/data\n",
    "ls *_fastqc* # list all fastqc files (which is in zip and html formats)\n",
    "\n",
    "mkdir fastqc # create fastqc folder\n",
    "mv *_fastqc* fastqc # move all fastqc files to fastqc folder"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a15d8a45-96e5-46cb-a8c1-7f0dd199ff89",
   "metadata": {},
   "source": [
    "# PRACTICAL EXERCISE 4\n",
    "Assignment: Run FASTQC on hcc1395_normal_1.fastq.gz you downloaded in the previous practical exercise."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a7edd69f-80e8-4c27-a071-1f5c200d055e",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Started analysis of hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep1_r1.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep1_r2.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep2_r1.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep2_r2.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep3_r1.fastq.gz\n",
      "Started analysis of hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Analysis complete for hcc1395_normal_rep3_r2.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep1_r1.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep1_r2.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep2_r1.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep2_r2.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep3_r1.fastq.gz\n",
      "Started analysis of hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 5% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 10% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 15% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 20% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 25% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 30% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 35% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 40% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 45% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 50% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 55% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 60% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 65% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 70% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 75% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 80% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 85% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 90% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Approx 95% complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "Analysis complete for hcc1395_tumor_rep3_r2.fastq.gz\n",
      "\n",
      "\u001b[38;5;208m///\u001b[0m \u001b]8;id=326166;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2mv1.28\u001b[0m\n",
      "\n",
      "\u001b[34m       file_search\u001b[0m | Search path: /workspace/rnaseq/practice/data\n",
      "\u001b[2K         \u001b[34msearching\u001b[0m | \u001b[38;5;70m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m37/37\u001b[0m  2m34/37\u001b[0m \u001b[2mhcc1395_normal_rep3_r1_fastqc.html\u001b[0m\n",
      "\u001b[?25h\u001b[34m            fastqc\u001b[0m | Found 12 reports\n",
      "\u001b[34m     write_results\u001b[0m | Data        : multiqc_data\n",
      "\u001b[34m     write_results\u001b[0m | Report      : multiqc_report.html\n",
      "\u001b[34m           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "cd $RNA_HOME/practice/data\n",
    "#fastqc hcc1395_normal_1.fastq.gz\n",
    "fastqc *.fastq.gz\n",
    "multiqc ./"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3fc6e52-efe0-449d-9b32-5bfe8048f7f3",
   "metadata": {},
   "source": [
    "* Then, go to the following url in your browser: http://YOUR_PUBLIC_IPv4_ADDRESS/rnaseq/practice/data/\n",
    "* fastqc for hcc1395_normal_rep1_r1: http://YOUR_PUBLIC_IPv4_ADDRESS/rnaseq/practice/data/hcc1395_normal_rep1_r1_fastqc.html\n",
    "\n",
    "You can see the total number of sequence in \"Basic statistics\" table at the top of fastqc report. When looking at Adaptor Content, there is some evidence of Illumina Universal Adapter at the 3’ ends of reads. This suggests that adapter trimming might be advisable for this data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e8f5f7f-fe09-4234-8dd8-d19d1a967363",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
